File: workloads/Haskell/RBT/src/Spec.hs
module Spec where

import Etna.Lib
import Data.Either
import qualified Data.List as L
import Impl

-- Based on SmallCheck implementation (based on Okasaki 1999),
-- but in the style of How to Specify It.

isRBT :: Ord k => Tree k v -> Bool
isRBT t = isBST t && consistentBlackHeight t && noRedRed t && blackRoot t

isBST :: Ord k => Tree k v -> Bool
isBST E = True
isBST (T _ a x _ b) =
  -- Difference from SC: don't allow repeated keys.
  every (< x) a && every (> x) b && isBST a && isBST b
  where
    every p E = True
    every p (T _ a x _ b) = p x && every p a && every p b

-- "No red node has a red parent."
noRedRed :: Tree k v -> Bool
noRedRed E = True
noRedRed (T B a _ _ b) = noRedRed a && noRedRed b
noRedRed (T R a _ _ b) = blackRoot a && blackRoot b && noRedRed a && noRedRed b
  where
    blackRoot (T R _ _ _ _) = False
    blackRoot _ = True

-- "Every path from the root to an empty node contains the same number of black nodes."
consistentBlackHeight :: Tree k v -> Bool
consistentBlackHeight = fst . go
  where
    go E = (True, 1)
    go (T rb a x _ b) =
      (aBool && bBool && aHeight == bHeight, aHeight + isBlack rb)
      where
        (aBool, aHeight) = go a
        (bBool, bHeight) = go b

        isBlack R = 0
        isBlack B = 1

blackRoot :: Tree k v -> Bool
blackRoot E = True
blackRoot (T B _ _ _ _) = True
blackRoot _ = False

toList :: Tree k v -> [(k, v)]
toList E = []
toList (T _ l k v r) =
  toList l ++ [(k, v)] ++ toList r

find :: Ord k => k -> Tree k v -> Maybe v
find _ E = Nothing
find x (T _ l y vy r)
  | x < y = find x l
  | x > y = find x r
  | otherwise = Just vy

----------

newtype Key = Key Int
  deriving (Eq, Ord, Read, Show)

newtype Val = Val Bool
  deriving (Eq, Ord, Read, Show)

type RBT = Tree Key Val

----------

-- Validity properties.

prop_InsertValid :: Task (RBT, Key, Val)
prop_InsertValid (t, k, v) =
  isRBT t --> fromRight False (isRBT <$> insert k v t)

prop_DeleteValid :: Task (RBT, Key)
prop_DeleteValid (t, k) =
  isRBT t --> fromRight False (isRBT <$> delete k t)

----------

-- Postcondition properties.

prop_InsertPost :: Task (RBT, Key, Key, Val)
prop_InsertPost (t, k, k', v) =
  isRBT t
    --> (find k' <$> insert k v t)
    == return (if k == k' then Just v else find k' t)

prop_DeletePost :: Task (RBT, Key, Key)
prop_DeletePost (t, k, k') =
  isRBT t
    --> (find k' <$> delete k t)
    == return (if k == k' then Nothing else find k' t)

----------

-- Model-based properties.

prop_InsertModel :: Task (RBT, Key, Val)
prop_InsertModel (t, k, v) =
  isRBT t
    --> (toList <$> insert k v t)
    == return (L.insert (k, v) (deleteKey k $ toList t))

prop_DeleteModel :: Task (RBT, Key)
prop_DeleteModel (t, k) =
  isRBT t
    --> (toList <$> delete k t)
    == return (deleteKey k (toList t))

deleteKey :: Ord k => k -> [(k, v)] -> [(k, v)]
deleteKey k = filter ((/= k) . fst)

----------

-- Metamorphic properties.

prop_InsertInsert :: Task (RBT, Key, Key, Val, Val)
prop_InsertInsert (t, k, k', v, v') =
  isRBT t
    --> (insert k v =<< insert k' v' t)
    =~= if k == k' then insert k v t else insert k' v' =<< insert k v t

prop_InsertDelete :: Task (RBT, Key, Key, Val)
prop_InsertDelete (t, k, k', v) =
  isRBT t
    --> (insert k v =<< delete k' t)
    =~= if k == k' then insert k v t else delete k' =<< insert k v t

prop_DeleteInsert :: Task (RBT, Key, Key, Val)
prop_DeleteInsert (t, k, k', v') =
  isRBT t
    --> (delete k =<< insert k' v' t)
    =~= if k == k' then delete k t else insert k' v' =<< delete k t

prop_DeleteDelete :: Task (RBT, Key, Key)
prop_DeleteDelete (t, k, k') =
  isRBT t
    --> (delete k =<< delete k' t)
    =~= (delete k' =<< delete k t)

(=~=) :: Either Error (Tree Key Val) -> Either Error (Tree Key Val) -> Bool
(Right t1) =~= (Right t2) = toList t1 == toList t2
_ =~= _ = False

----------

sizeRBT :: RBT -> Int
sizeRBT = length . toList

File: workloads/OCaml/RBT/lib/spec.ml
open Impl
open QCheck

let rec isBST (t : ('a, 'b) tree) : bool =
  let rec every (p : 'a -> bool) (t : ('a, 'b) tree) : bool =
    match t with
    | E -> true
    | T (_, a, x, _, b) -> p x && every p a && every p b
  in
  match t with
  | E -> true
  | T (_, a, x, _, b) ->
      every (( > ) x) a && every (( < ) x) b && isBST a && isBST b

let rec noRedRed (t : ('a, 'b) tree) : bool =
  let blackRoot (t : ('a, 'b) tree) : bool =
    match t with T (R, _, _, _, _) -> false | _ -> true
  in
  match t with
  | E -> true
  | T (B, a, _, _, b) -> noRedRed a && noRedRed b
  | T (R, a, _, _, b) -> blackRoot a && blackRoot b && noRedRed a && noRedRed b

let consistentBlackHeight (t : ('a, 'b) tree) : bool =
  let rec go (t : ('a, 'b) tree) : bool * int =
    match t with
    | E -> (true, 1)
    | T (rb, a, _, _, b) ->
        let aBool, aHeight = go a in
        let bBool, bHeight = go b in
        let isBlack (rb : color) : int = match rb with B -> 1 | R -> 0 in
        (aBool && bBool && aHeight = bHeight, aHeight + isBlack rb)
  in
  fst (go t)

let isRBT (t : ('a, 'b) tree) : bool =
  isBST t && consistentBlackHeight t && noRedRed t

let rec toList (t : ('a, 'b) tree) : ('a * 'b) list =
  match t with E -> [] | T (_, l, k, v, r) -> toList l @ [ (k, v) ] @ toList r

(* -- Validity properties. *)

let prop_InsertValid (t : ('a, 'b) tree) (k : 'a) (v : 'b) =
  assume (isRBT t);
  isRBT (insert k v t)

let prop_DeleteValid (t : ('a, 'b) tree) (k : 'a) =
  assume (isRBT t);
  delete k t >>= fun t' -> Some (isRBT t')

(* ---------- *)

let prop_InsertPost (t : ('a, 'b) tree) (k : 'a) (k' : 'a) (v : 'b) : bool =
  assume (isRBT t);
  let v' = find k' (insert k v t) in
  if k = k' then v' = Some v else v' = find k' t

let prop_DeletePost (t : ('a, 'b) tree) (k : 'a) (k' : 'a) : bool =
  assume (isRBT t);
  delete k t >>= (fun t' -> find k' t') = if k = k' then None else find k' t

(* ---------- *)

(* -- Model-based properties. *)

let deleteKey (k : 'a) (l : ('a * 'b) list) : ('a * 'b) list =
  let rec filter f l =
    match l with
    | [] -> []
    | x :: xs -> if f x then x :: filter f xs else filter f xs
  in
  filter (fun x -> not (fst x = k)) l

let rec l_insert (kv : 'a * 'b) (l : ('a * 'b) list) : ('a * 'b) list =
  match l with
  | [] -> [ kv ]
  | (k, v) :: xs ->
      if fst kv = k then kv :: xs
      else if fst kv < k then kv :: l
      else (k, v) :: l_insert kv xs

let prop_InsertModel (t : ('a, 'b) tree) (k : 'a) (v : 'b) =
  assume (isRBT t);
  toList (insert k v t) = l_insert (k, v) (deleteKey k (toList t))

let prop_DeleteModel (t : ('a, 'b) tree) (k : 'a) =
  assume (isRBT t);
  delete k t >>= fun t' -> Some (toList t' = deleteKey k (toList t))

(* ---------- *)

(* -- Metamorphic properties. *)

let prop_InsertInsert (t : ('a, 'b) tree) (k : 'a) (k' : 'a) (v : 'b) (v' : 'b)
    : bool =
  assume (isRBT t);
  toList (insert k v (insert k' v' t))
  = toList (if k = k' then insert k v t else insert k' v' (insert k v t))

let prop_InsertDelete (t : ('a, 'b) tree) (k : 'a) (k' : 'a) (v : 'b) :
    bool option =
  assume (isRBT t);
  delete k' t >>= fun t' ->
  delete k' (insert k v t) >>= fun t'' ->
  Some (toList (insert k v t') = toList (if k = k' then insert k v t else t''))

let prop_DeleteInsert (t : ('a, 'b) tree) (k : 'a) (k' : 'a) (v' : 'b) =
  assume (isRBT t);
  delete k (insert k' v' t) >>= fun t' ->
  delete k t >>= fun t'' ->
  let t''' = insert k' v' t'' in
  Some (toList t' = toList (if k = k' then t'' else t'''))

let prop_DeleteDelete (t : ('a, 'b) tree) (k : 'a) (k' : 'a) =
  assume (isRBT t);
  delete k' t >>= fun t' ->
  delete k t' >>= fun t'' ->
  delete k t >>= fun t1' ->
  delete k' t1' >>= fun t1'' -> Some (toList t'' = toList t1'')

(* ---------- *)

let sizeRBT (t : ('a, 'b) tree) : int =
  let rec length l = match l with [] -> 0 | _ :: xs -> 1 + length xs in
  length (toList t)
